errors while running integration tests locally

I'm trying to run integration tests locally and running into a bunch of errors. I think it'll be better to run the integration tests in the CI in the same way that we're expecting them to be executed locally; otherwise the local ones might break and we won't know.

```
============================================================================== short test summary info ==============================================================================
FAILED src/tests/integration/test_containers.py::test_invidual_container[postgres-database system is ready to accept connections-postgreSQL] - AssertionError: assert False
FAILED src/tests/integration/test_containers.py::test_invidual_container[mysql-mysqld: ready for connections-mySQL] - AssertionError: assert False
FAILED src/tests/integration/test_generic_db_opeations.py::test_query_count[ip_with_postgreSQL-3] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_query_count[ip_with_mySQL-3] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_create_table_with_indexed_df[ip_with_postgreSQL-15] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_create_table_with_indexed_df[ip_with_mySQL-15] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_active_connection_number[ip_with_postgreSQL-1] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_active_connection_number[ip_with_mySQL-1] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_close_and_connect[ip_with_postgreSQL-postgreSQL] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_close_and_connect[ip_with_mySQL-mySQL] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_telemetry_execute_command_has_connection_info[ip_with_postgreSQL-postgresql-psycopg2] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_telemetry_execute_command_has_connection_info[ip_with_mySQL-mysql-pymysql] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_histogram[ip_with_postgreSQL-histogram] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_histogram[ip_with_postgreSQL-hist] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_histogram[ip_with_postgreSQL-histogram-bins] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_histogram[ip_with_mySQL-histogram] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_histogram[ip_with_mySQL-hist] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_histogram[ip_with_mySQL-histogram-bins] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_boxplot[ip_with_postgreSQL-boxplot] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_boxplot[ip_with_postgreSQL-box] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_boxplot[ip_with_postgreSQL-boxplot-horizontal] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_sqlplot_boxplot[ip_with_postgreSQL-boxplot-with] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_sql_cmd_magic_uno[ip_with_postgreSQL] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_sql_cmd_magic_uno[ip_with_mySQL] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_sql_cmd_magic_dos[ip_with_postgreSQL] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
FAILED src/tests/integration/test_generic_db_opeations.py::test_sql_cmd_magic_dos[ip_with_mySQL] - sqlalchemy.exc.OperationalError: (pymysql.err.OperationalError) (2003, "Can't connect to MySQL server on 'localhost' ([Errno 61] Connection refused)")
FAILED src/tests/integration/test_generic_db_opeations.py::test_profile_query[ip_with_postgreSQL-taxi-table_columns0-expected0] - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
ERROR src/tests/integration/test_postgreSQL.py::test_meta_cmd_display - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
ERROR src/tests/integration/test_postgreSQL.py::test_auto_commit_mode_on - sqlalchemy.exc.OperationalError: (psycopg2.OperationalError) could not connect to server: Connection refused
```

--- a/.github/workflows/ci-integration-db.yaml
+++ b/.github/workflows/ci-integration-db.yaml
@@ -23,4 +23,4 @@ jobs:
       - name: Integration Test
         run: |
           # Run the integration test by pytest marker
-          pytest src/tests/integration
+          pytest src/tests/integration --verbose

--- a/.github/workflows/ci-integration-db.yaml
+++ b/.github/workflows/ci-integration-db.yaml
@@ -9,58 +9,6 @@ jobs:
         python-version: ['3.11']
         os: [ubuntu-latest]
     runs-on: ${{ matrix.os }}
-    services:
-    # postgreSQL Service
-      postgres:
-        image: postgres
-        env:
-          POSTGRES_DB: db
-          POSTGRES_USER: ploomber_app
-          POSTGRES_PASSWORD: ploomber_app_password
-        options: >-
-          --health-cmd pg_isready
-          --health-interval 10s
-          --health-timeout 5s
-          --health-retries 5
-        ports:
-          - 5432:5432
-      # mySQL Service
-      mysql:
-        image: mysql
-        env:
-          MYSQL_DATABASE: db
-          MYSQL_USER: ploomber_app
-          MYSQL_PASSWORD: ploomber_app_password
-          MYSQL_ROOT_PASSWORD: ploomber_app_root_password
-        ports:
-          - 33306:3306
-        options: --health-cmd="mysqladmin ping" --health-interval=10s --health-timeout=5s --health-retries=3
-      mariadb:
-      # mariaDB Service
-        image: mariadb
-        env:
-          MYSQL_DATABASE: db
-          MYSQL_USER: ploomber_app
-          MYSQL_PASSWORD: ploomber_app_password
-          MYSQL_ROOT_PASSWORD: ploomber_app_root_password
-        ports:
-          - 33309:3306
-        options: --health-cmd="mysqladmin ping" --health-interval=5s --health-timeout=2s --health-retries=3
-      # MSSQL Service
-      mssql:
-        image: mcr.microsoft.com/azure-sql-edge
-        env:
-          ACCEPT_EULA: Y
-          MSSQL_DATABASE: master
-          MSSQL_USER: sa
-          MSSQL_SA_PASSWORD: Ploomber_App@Root_Password
-        ports: 
-          - 1433:1433
-        options: >-
-            --health-cmd "/opt/mssql-tools/bin/sqlcmd -U $DB_USER -P $SA_PASSWORD -Q 'select 1' -b -o /dev/null"
-            --health-interval 5s
-            --health-timeout 2s
-            --health-retries 3
     steps:
       - uses: actions/checkout@v2
 

--- a/.github/workflows/ci-integration-db.yaml
+++ b/.github/workflows/ci-integration-db.yaml
@@ -19,15 +19,6 @@ jobs:
 
       - name: Install dependencies
         run: |
-          # Install MSSQL ODBC 18
-          curl https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -
-          sudo curl -o /etc/apt/sources.list.d/mssql-release.list https://packages.microsoft.com/config/ubuntu/$(lsb_release -rs)/prod.list 
-          sudo apt-get update
-          sudo ACCEPT_EULA=Y apt-get install -y msodbcsql18
-          sudo ACCEPT_EULA=Y apt-get install -y mssql-tools18
-          echo 'export PATH="$PATH:/opt/mssql-tools18/bin"' >> ~/.bashrc
-          source ~/.bashrc
-
           pip install ".[integration]"
       - name: Integration Test
         run: |

--- a/.github/workflows/ci-integration-db.yaml
+++ b/.github/workflows/ci-integration-db.yaml
@@ -19,6 +19,15 @@ jobs:
 
       - name: Install dependencies
         run: |
+          # Install MSSQL ODBC 18
+          curl https://packages.microsoft.com/keys/microsoft.asc | sudo apt-key add -
+          sudo curl -o /etc/apt/sources.list.d/mssql-release.list https://packages.microsoft.com/config/ubuntu/$(lsb_release -rs)/prod.list 
+          sudo apt-get update
+          sudo ACCEPT_EULA=Y apt-get install -y msodbcsql18
+          sudo ACCEPT_EULA=Y apt-get install -y mssql-tools18
+          echo 'export PATH="$PATH:/opt/mssql-tools18/bin"' >> ~/.bashrc
+          source ~/.bashrc
+
           pip install ".[integration]"
       - name: Integration Test
         run: |

--- a/src/sql/_testing.py
+++ b/src/sql/_testing.py
@@ -132,13 +132,10 @@ client = docker.from_env()
 def database_ready(
     database,
     timeout=20,
-    poll_freq=0.2,
+    poll_freq=0.5,
 ):
-    """Wait until a postgres instance is ready to receive connections.
+    """Wait until the container is ready to receive connections.
 
-    .. note::
-
-        This requires psycopg2 to be installed.
 
     :type host: str
     :type port: int
@@ -147,15 +144,23 @@ def database_ready(
     """
     import sqlalchemy
 
+    errors = []
+
     t0 = time.time()
     while time.time() - t0 < timeout:
         try:
             eng = sqlalchemy.create_engine(_get_database_url(database)).connect()
             eng.close()
             return True
-        except Exception:
-            pass
+        except Exception as e:
+            errors.append(str(e))
+
         time.sleep(poll_freq)
+
+    # print all the errors so we know what's goin on since failing to connect might be
+    # to some misconfiguration error
+    errors_ = "\n".join(errors)
+    print(f"ERRORS: {errors_}")
 
     return False
 

--- a/src/tests/integration/conftest.py
+++ b/src/tests/integration/conftest.py
@@ -1,4 +1,3 @@
-import os
 import shutil
 import pandas as pd
 import pytest
@@ -8,7 +7,6 @@ from sql import _testing
 is_on_github = False
 # if "GITHUB_ACTIONS" in os.environ:
 #     is_on_github = True
-
 
 
 @pytest.fixture

--- a/src/tests/integration/conftest.py
+++ b/src/tests/integration/conftest.py
@@ -6,8 +6,9 @@ from sqlalchemy import create_engine
 from sql import _testing
 
 is_on_github = False
-if "GITHUB_ACTIONS" in os.environ:
-    is_on_github = True
+# if "GITHUB_ACTIONS" in os.environ:
+#     is_on_github = True
+
 
 
 @pytest.fixture

--- a/src/tests/integration/conftest.py
+++ b/src/tests/integration/conftest.py
@@ -4,10 +4,6 @@ import pytest
 from sqlalchemy import create_engine
 from sql import _testing
 
-is_on_github = False
-# if "GITHUB_ACTIONS" in os.environ:
-#     is_on_github = True
-
 
 @pytest.fixture
 def get_database_config_helper():
@@ -50,7 +46,7 @@ def load_numeric_data(engine):
 
 @pytest.fixture(scope="session")
 def setup_postgreSQL():
-    with _testing.postgres(is_bypass_init=is_on_github):
+    with _testing.postgres():
         engine = create_engine(
             _testing.DatabaseConfigHelper.get_database_url("postgreSQL")
         )
@@ -81,7 +77,7 @@ def ip_with_postgreSQL(ip_empty, setup_postgreSQL):
 
 @pytest.fixture(scope="session")
 def setup_mySQL():
-    with _testing.mysql(is_bypass_init=is_on_github):
+    with _testing.mysql():
         engine = create_engine(_testing.DatabaseConfigHelper.get_database_url("mySQL"))
         # Load pre-defined datasets
         load_taxi_data(engine)
@@ -110,7 +106,7 @@ def ip_with_mySQL(ip_empty, setup_mySQL):
 
 @pytest.fixture(scope="session")
 def setup_mariaDB():
-    with _testing.mariadb(is_bypass_init=is_on_github):
+    with _testing.mariadb():
         engine = create_engine(
             _testing.DatabaseConfigHelper.get_database_url("mariaDB")
         )
@@ -197,7 +193,7 @@ def ip_with_duckDB(ip_empty, setup_duckDB):
 
 @pytest.fixture(scope="session")
 def setup_MSSQL():
-    with _testing.mssql(is_bypass_init=is_on_github):
+    with _testing.mssql():
         engine = create_engine(_testing.DatabaseConfigHelper.get_database_url("MSSQL"))
         # Load taxi_data
         load_taxi_data(engine)

--- a/src/tests/integration/test_generic_db_operations.py
+++ b/src/tests/integration/test_generic_db_operations.py
@@ -414,3 +414,26 @@ def test_profile_query(request, ip_with_dynamic_db, table, table_columns, expect
 
             assert criteria in expected
             assert cell_value == str(expected[criteria][i])
+
+
+@pytest.mark.parametrize(
+    "cell",
+    [
+        "%sqlcmd tables",
+        "%sqlcmd columns --table numbers",
+    ],
+)
+@pytest.mark.parametrize(
+    "ip_with_dynamic_db",
+    [
+        ("ip_with_postgreSQL"),
+        ("ip_with_mySQL"),
+        ("ip_with_mariaDB"),
+        ("ip_with_SQLite"),
+        ("ip_with_duckDB"),
+    ],
+)
+def test_sqlcmd_tables_columns(ip_with_dynamic_db, cell, request):
+    ip_with_dynamic_db = request.getfixturevalue(ip_with_dynamic_db)
+    out = ip_with_dynamic_db.run_cell(cell)
+    assert out.result

--- a/src/tests/integration/test_generic_db_operations.py
+++ b/src/tests/integration/test_generic_db_operations.py
@@ -6,6 +6,15 @@ from sql.telemetry import telemetry
 from unittest.mock import ANY, Mock
 import math
 
+ALL_DATABASES = [
+    "ip_with_postgreSQL",
+    "ip_with_mySQL",
+    "ip_with_mariaDB",
+    "ip_with_SQLite",
+    "ip_with_duckDB",
+    "ip_with_MSSQL",
+]
+
 
 @pytest.fixture(autouse=True)
 def run_around_tests(tmpdir_factory):
@@ -249,17 +258,7 @@ def test_sqlplot_boxplot(ip_with_dynamic_db, cell, request):
     assert type(out.result).__name__ in {"Axes", "AxesSubplot"}
 
 
-@pytest.mark.parametrize(
-    "ip_with_dynamic_db",
-    [
-        ("ip_with_postgreSQL"),
-        ("ip_with_mySQL"),
-        ("ip_with_mariaDB"),
-        ("ip_with_SQLite"),
-        ("ip_with_duckDB"),
-        ("ip_with_MSSQL"),
-    ],
-)
+@pytest.mark.parametrize("ip_with_dynamic_db", ALL_DATABASES)
 def test_sql_cmd_magic_uno(ip_with_dynamic_db, request):
     ip_with_dynamic_db = request.getfixturevalue(ip_with_dynamic_db)
 
@@ -273,17 +272,7 @@ def test_sql_cmd_magic_uno(ip_with_dynamic_db, request):
     assert "greater" in result.keys()
 
 
-@pytest.mark.parametrize(
-    "ip_with_dynamic_db",
-    [
-        ("ip_with_postgreSQL"),
-        ("ip_with_mySQL"),
-        ("ip_with_mariaDB"),
-        ("ip_with_SQLite"),
-        ("ip_with_duckDB"),
-        ("ip_with_MSSQL"),
-    ],
-)
+@pytest.mark.parametrize("ip_with_dynamic_db", ALL_DATABASES)
 def test_sql_cmd_magic_dos(ip_with_dynamic_db, request):
     ip_with_dynamic_db = request.getfixturevalue(ip_with_dynamic_db)
 
@@ -423,17 +412,21 @@ def test_profile_query(request, ip_with_dynamic_db, table, table_columns, expect
         "%sqlcmd columns --table numbers",
     ],
 )
-@pytest.mark.parametrize(
-    "ip_with_dynamic_db",
-    [
-        ("ip_with_postgreSQL"),
-        ("ip_with_mySQL"),
-        ("ip_with_mariaDB"),
-        ("ip_with_SQLite"),
-        ("ip_with_duckDB"),
-    ],
-)
+@pytest.mark.parametrize("ip_with_dynamic_db", ALL_DATABASES)
 def test_sqlcmd_tables_columns(ip_with_dynamic_db, cell, request):
     ip_with_dynamic_db = request.getfixturevalue(ip_with_dynamic_db)
     out = ip_with_dynamic_db.run_cell(cell)
     assert out.result
+
+
+@pytest.mark.parametrize(
+    "cell",
+    [
+        "%%sql\nSELECT *\n-- %variable\nFROM NUMBERS WHERE 0=1",
+    ],
+)
+@pytest.mark.parametrize("ip_with_dynamic_db", ALL_DATABASES)
+def test_sql_query_with_interpolation_like_comments(ip_with_dynamic_db, cell, request):
+    ip_with_dynamic_db = request.getfixturevalue(ip_with_dynamic_db)
+    out = ip_with_dynamic_db.run_cell(cell)
+    assert out.error_in_exec is None

--- a/src/tests/integration/test_generic_db_operations.py
+++ b/src/tests/integration/test_generic_db_operations.py
@@ -191,6 +191,10 @@ def test_telemetry_execute_command_has_connection_info(
         ("ip_with_mariaDB"),
         ("ip_with_SQLite"),
         ("ip_with_duckDB"),
+        pytest.param(
+            "ip_with_MSSQL",
+            marks=pytest.mark.xfail(reason="sqlglot does not support SQL server"),
+        ),
     ],
 )
 def test_sqlplot_histogram(ip_with_dynamic_db, cell, request):
@@ -422,11 +426,23 @@ def test_sqlcmd_tables_columns(ip_with_dynamic_db, cell, request):
 @pytest.mark.parametrize(
     "cell",
     [
-        "%%sql\nSELECT *\n-- %variable\nFROM NUMBERS WHERE 0=1",
+        "%%sql\nSELECT * FROM numbers WHERE 0=1",
+        "%%sql --with subset\nSELECT * FROM subset WHERE 0=1",
+        "%%sql\nSELECT *\n-- %one $another\nFROM numbers WHERE 0=1",
+    ],
+    ids=[
+        "simple-query",
+        "cte",
+        "interpolation-like-comment",
     ],
 )
 @pytest.mark.parametrize("ip_with_dynamic_db", ALL_DATABASES)
-def test_sql_query_with_interpolation_like_comments(ip_with_dynamic_db, cell, request):
+def test_sql_query(ip_with_dynamic_db, cell, request):
     ip_with_dynamic_db = request.getfixturevalue(ip_with_dynamic_db)
+    ip_with_dynamic_db.run_cell(
+        """%%sql --save subset --no-execute
+SELECT * FROM numbers WHERE 1=0
+"""
+    )
     out = ip_with_dynamic_db.run_cell(cell)
     assert out.error_in_exec is None

--- a/src/tests/integration/test_mssql.py
+++ b/src/tests/integration/test_mssql.py
@@ -16,7 +16,6 @@ def test_query_count(ip_with_MSSQL):
     assert len(out) == 3
 
 
-@pytest.mark.xfail(reason="Known sqlglot issue, addressing in: jupysql/issues/307")
 def test_cte(ip_with_MSSQL):
     ip_with_MSSQL.run_cell(
         "%sql --save taxi_subset --no-execute \

